ocaml/libs/tracing/tracing.ml

```diff
@@ -576,17 +576,70 @@ module Export = struct
 
       let set_trace_log_dir dir = trace_log_dir := dir
 
-      let export ~trace_id ~span_json ~path : (string, exn) result =
-        try
+      module FileStream = struct
```

Comment:
Because we don't want to leave the file descriptrs only we really want to expose an interface to the Filestream that will close the resources implicitly when the user function finishes writing everything.

This can be done with a Filestream.with function that has a function as a parameter that is given an export function which can use. and once this function finishes, the finaliser can close the file descriptors.

```ocaml
module Filestream = struct
  val with : unit -> (span_json : ... -> (string, exn) result) -> unit) -> unit
end
...
    let export_to_endpoint span_list endpoint =
      try
        debug "Tracing: About to export" ;
        let ( let@ ) f x = f x in
        let@ file_export = Filestream.(with ~dir:!trace_log_dir ()) in
        let export = match endpoint with
        | Url url ->  Http.export ~url
        | Bugtool -> file_export
        Hashtbl.iter (...) span_list
      with e -> debug "Tracing: ERROR %s" (Printexc.to_string e)
```

Once this interface is stablished I think the module File doesn't contain anything as the export function is not provided by Filestream, and can be dropped. This design also allows set_trace_log_dir to be stable over a whole run of span exports, previously it could be changed while the export was in process.

Reply:
Part of the idea currently is to keep resources open throughout the iter, would this be affect here?